File: lab6-openmp/lab6-openmp_f90/omp_workshare1.f90
! -*- Mode: Fortran90; -*-
!=================================================================

program WorkShare1
  !-----------------------------------------------------------------
  ! FILE: omp_workshare1.f90
  ! DESCRIPTION:
  !   OpenMP Example - Loop Work-sharing 
  !   In this example, the iterations of a loop are scheduled 
  !   dynamically across the team of threads.  A thread will 
  !   perform chunk iterations at a time before being scheduled 
  !   for the next chunk of work.
  ! AUTHOR: Blaise Barney  5/99, 1/09/04
  ! Upgraded to F90, free-source form: Daniel R. Reynolds, 2/2009
  !-----------------------------------------------------------------
  !======= Inclusions ===========

  !======= Declarations =========
  implicit none

  integer :: NThreads, TID, chunk, I 
  integer, external :: omp_get_num_threads, omp_get_thread_num
  integer, parameter :: N = 100
  integer, parameter :: chunksize = 10
  real :: A(N), B(N), C(N)

  !======= Internals ============

  ! Some initializations
  do I = 1, N
     A(I) = I * 1.0
     B(I) = A(I)
  enddo
  chunk = chunksize

  !$omp parallel shared(A,B,C,NThreads,chunk) private(I,TID)

  TID = omp_get_thread_num()
  if (TID .eq. 0) then
     NThreads = omp_get_num_threads()
     print *, 'Number of threads =', NThreads
  end if
  print *, 'Thread',TID,' starting...'

  !$omp do schedule(dynamic,chunk)
  do I = 1, N
     C(I) = A(I) + B(I)
     print *, ' Thread',TID,': C(',I,')=',C(I)
  enddo
  !$omp end do nowait

  print *, 'Thread',TID,' done.'

  !$omp end parallel

end program WorkShare1
!=================================================================

File: lab6-openmp/lab6-openmp_cxx/omp_workshare1.cpp
/* FILE: omp_workshare1.cpp
   DESCRIPTION:
      OpenMP Example - Loop Work-sharing - C++ Version
      In this example, the iterations of a loop are scheduled dynamically
      across the team of threads.  A thread will perform CHUNK iterations
      at a time before being scheduled for the next CHUNK of work.
   AUTHOR: Blaise Barney  5/99
   UPDATED: Daniel R. Reynolds (updated to C++), 1/13/2013 */

#include <omp.h>
#include <stdio.h>
#include <stdlib.h>
#define CHUNKSIZE  10
#define N          100


int main (int argc, char *argv[]) {

  // local variables
  int nthreads, tid, i, chunk;
  float a[N], b[N], c[N];

  // Some initializations
  for (i=0; i < N; i++)  a[i] = b[i] = i * 1.0;
  chunk = CHUNKSIZE;

# pragma omp parallel shared(a,b,c,nthreads,chunk) private(i,tid)
  {
    tid = omp_get_thread_num();
    if (tid == 0) {
      nthreads = omp_get_num_threads();
      printf("Number of threads = %i\n", nthreads);
    }
    printf("Thread %i starting...\n", tid);

#   pragma omp for schedule(dynamic,chunk)
    for (i=0; i<N; i++) {
      c[i] = a[i] + b[i];
      printf("Thread %i: c[%i]= %g\n", tid, i, c[i]);
    }

  }  // end parallel 

  return 0;
}  // end main
